## Data Pre-processing

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
ipk = pd.Series([3.14,3.33,2.8,2.4,np.nan,3.24,3,3.99,3.33])

In [ ]:
ipk

0    3.14
1    3.33
2    2.80
3    2.40
4     NaN
5    3.24
6    3.00
7    3.99
8    3.33
dtype: float64

In [ ]:
nama = pd.Series(["A","B","C","D","E","F","G","H","B"])
gender = pd.Series(["Female","M","Male","F","M",np.nan,"F","F","M"])

In [ ]:
data={'nama': nama,
     'ipk': ipk,
     'gender': gender}

In [ ]:
mahasiswa = pd.DataFrame(data)
mahasiswa

nama   ipk  gender
0    A  3.14  Female
1    B  3.33       M
2    C  2.80    Male
3    D  2.40       F
4    E   NaN       M
5    F  3.24     NaN
6    G  3.00       F
7    H  3.99       F
8    B  3.33       M

In [ ]:
new_data=mahasiswa.drop_duplicates()
new_data

nama   ipk  gender
0    A  3.14  Female
1    B  3.33       M
2    C  2.80    Male
3    D  2.40       F
4    E   NaN       M
5    F  3.24     NaN
6    G  3.00       F
7    H  3.99       F

In [ ]:
mahasiswa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 3 columns):
nama      9 non-null object
ipk       8 non-null float64
gender    8 non-null object
dtypes: float64(1), object(2)
memory usage: 344.0+ bytes


In [ ]:
new_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8 entries, 0 to 7
Data columns (total 3 columns):
nama      8 non-null object
ipk       7 non-null float64
gender    7 non-null object
dtypes: float64(1), object(2)
memory usage: 256.0+ bytes


In [ ]:
new_data1=new_data.dropna()
new_data1

nama   ipk  gender
0    A  3.14  Female
1    B  3.33       M
2    C  2.80    Male
3    D  2.40       F
6    G  3.00       F
7    H  3.99       F

In [ ]:
new_data1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6 entries, 0 to 7
Data columns (total 3 columns):
nama      6 non-null object
ipk       6 non-null float64
gender    6 non-null object
dtypes: float64(1), object(2)
memory usage: 192.0+ bytes


In [ ]:
rata_ipk=new_data['ipk'].mean()
rata_ipk

3.1285714285714286

In [ ]:
new_data2 = new_data['ipk'].fillna(rata_ipk)
new_data2

0    3.140000
1    3.330000
2    2.800000
3    2.400000
4    3.128571
5    3.240000
6    3.000000
7    3.990000
Name: ipk, dtype: float64

In [ ]:
new_data['ipk']=new_data['ipk'].fillna(rata_ipk)

In [ ]:
new_data

nama       ipk gender
0    A  3.140000      F
1    B  3.330000      M
2    C  2.800000      M
3    D  2.400000      F
4    E  3.128571      M
5    F  3.240000      M
6    G  3.000000      F
7    H  3.990000      F

In [ ]:
new_data=new_data.replace({'gender':{'Male': 'M', 'Female':'F'}})

In [ ]:
new_data

nama       ipk gender
0    A  3.140000      F
1    B  3.330000      M
2    C  2.800000      M
3    D  2.400000      F
4    E  3.128571      M
5    F  3.240000      M
6    G  3.000000      F
7    H  3.990000      F

In [ ]:
modus=new_data['gender'].value_counts().index[0]
modus

'F'

In [ ]:
new_data['gender']=new_data['gender'].fillna(modus)
new_data

nama       ipk gender
0    A  3.140000      F
1    B  3.330000      M
2    C  2.800000      M
3    D  2.400000      F
4    E  3.128571      M
5    F  3.240000      M
6    G  3.000000      F
7    H  3.990000      F

# Regresi Linier

In [ ]:
salary = pd.read_csv("salary_data.csv",sep=",")
salary.head()

YearsExperience   Salary
0              1.1  39343.0
1              1.3  46205.0
2              1.5  37731.0
3              2.0  43525.0
4              2.2  39891.0

In [ ]:
x = salary.iloc[:,0]
x.head()

0    1.1
1    1.3
2    1.5
3    2.0
4    2.2
Name: YearsExperience, dtype: float64

In [ ]:
y = salary.iloc[:,1]
y.head()

0    39343.0
1    46205.0
2    37731.0
3    43525.0
4    39891.0
Name: Salary, dtype: float64

In [ ]:
salary.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 2 columns):
YearsExperience    30 non-null float64
Salary             30 non-null float64
dtypes: float64(2)
memory usage: 608.0 bytes


In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.2,random_state=0) ##24 train dan 6 test

In [ ]:
len(x_train)

24

In [ ]:
x_train.head()

27    9.6
11    4.0
17    5.3
22    7.9
5     2.9
Name: YearsExperience, dtype: float64

In [ ]:
x_test

2      1.5
28    10.3
13     4.1
10     3.9
26     9.5
24     8.7
Name: YearsExperience, dtype: float64

In [ ]:
len(x_test)

6

In [ ]:
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [ ]:
data = {'x_tr':x_train,
       'y_tr':y_train}

In [ ]:
data_sales = pd.DataFrame(data)
data_sales.head()

x_tr      y_tr
27   9.6  112635.0
11   4.0   55794.0
17   5.3   83088.0
22   7.9  101302.0
5    2.9   56642.0

In [ ]:
model2 = smf.ols('y_tr~x_tr',data=data_sales).fit()
model2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   y_tr   R-squared:                       0.941
Model:                            OLS   Adj. R-squared:                  0.939
Method:                 Least Squares   F-statistic:                     352.1
Date:                Tue, 19 Mar 2024   Prob (F-statistic):           5.03e-15
Time:                        16:54:58   Log-Likelihood:                -242.89
No. Observations:                  24   AIC:                             489.8
Df Residuals:                      22   BIC:                             492.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   2.678e+04   2818.687      9.501      0.000    2.09e+04    3.26e+04
x_tr        9312.5751    496.279     18.765      0.000    8283.355    1.03e+04
==============================================================================
Omnibus:                        3.105   Durbin-Watson:                   2.608
Prob(Omnibus):                  0.212   Jarque-Bera (JB):                1.567
Skew:                           0.297   Prob(JB):                        0.457
Kurtosis:                       1.898   Cond. No.                         12.8
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [ ]:
startup = pd.read_csv("50_Startups.csv",sep = ",")
startup.head()

RD  Administration  Marketing       State     Profit
0  165349.20       136897.80  471784.10    New York  192261.83
1  162597.70       151377.59  443898.53  California  191792.06
2  153441.51       101145.55  407934.54     Florida  191050.39
3  144372.41       118671.85  383199.62    New York  182901.99
4  142107.34        91391.77  366168.42     Florida  166187.94

In [ ]:
model3 = smf.ols('Profit~RD+Administration+Marketing',data = startup).fit()

In [ ]:
model3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Profit   R-squared:                       0.951
Model:                            OLS   Adj. R-squared:                  0.948
Method:                 Least Squares   F-statistic:                     296.0
Date:                Tue, 19 Mar 2024   Prob (F-statistic):           4.53e-30
Time:                        16:55:04   Log-Likelihood:                -525.39
No. Observations:                  50   AIC:                             1059.
Df Residuals:                      46   BIC:                             1066.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept       5.012e+04   6572.353      7.626      0.000    3.69e+04    6.34e+04
RD                 0.8057      0.045     17.846      0.000       0.715       0.897
Administration    -0.0268      0.051     -0.526      0.602      -0.130       0.076
Marketing          0.0272      0.016      1.655      0.105      -0.006       0.060
==============================================================================
Omnibus:                       14.838   Durbin-Watson:                   1.282
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               21.442
Skew:                          -0.949   Prob(JB):                     2.21e-05
Kurtosis:                       5.586   Cond. No.                     1.40e+06
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### RD = RD mempengaruhi profit
#### Administration = Administration tidak mempengaruhi profit
#### Marketing = Marketing tidak mempengaruhi profit

In [ ]:
model4 = smf.ols('Profit~RD+Administration',data = startup).fit()
model4.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Profit   R-squared:                       0.948
Model:                            OLS   Adj. R-squared:                  0.946
Method:                 Least Squares   F-statistic:                     426.8
Date:                Tue, 19 Mar 2024   Prob (F-statistic):           7.29e-31
Time:                        16:55:06   Log-Likelihood:                -526.83
No. Observations:                  50   AIC:                             1060.
Df Residuals:                      47   BIC:                             1065.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept       5.489e+04   6016.718      9.122      0.000    4.28e+04     6.7e+04
RD                 0.8621      0.030     28.589      0.000       0.801       0.923
Administration    -0.0530      0.049     -1.073      0.289      -0.152       0.046
==============================================================================
Omnibus:                       14.678   Durbin-Watson:                   1.189
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               20.449
Skew:                          -0.961   Prob(JB):                     3.63e-05
Kurtosis:                       5.474   Cond. No.                     6.65e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.65e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### RD = RD mempengaruhi profit (alpha 0,1)
#### Administration = Administration tidak mempengaruhi profit

In [ ]:
model5 = smf.ols('Profit~RD+Marketing',data = startup).fit()
model5.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Profit   R-squared:                       0.950
Model:                            OLS   Adj. R-squared:                  0.948
Method:                 Least Squares   F-statistic:                     450.8
Date:                Tue, 19 Mar 2024   Prob (F-statistic):           2.16e-31
Time:                        16:55:08   Log-Likelihood:                -525.54
No. Observations:                  50   AIC:                             1057.
Df Residuals:                      47   BIC:                             1063.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   4.698e+04   2689.933     17.464      0.000    4.16e+04    5.24e+04
RD             0.7966      0.041     19.266      0.000       0.713       0.880
Marketing      0.0299      0.016      1.927      0.060      -0.001       0.061
==============================================================================
Omnibus:                       14.677   Durbin-Watson:                   1.257
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               21.161
Skew:                          -0.939   Prob(JB):                     2.54e-05
Kurtosis:                       5.575   Cond. No.                     5.32e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.32e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### RD mempengaruhi profit
#### Marketing mempengaruhi profit (alpha 0,1)
#### ADJ R^2 = 0.948
#### AIC = 1057
#### Model 5 paling bagus

In [ ]:
model6 = smf.ols('Profit~RD',data = startup).fit()
model6.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Profit   R-squared:                       0.947
Model:                            OLS   Adj. R-squared:                  0.945
Method:                 Least Squares   F-statistic:                     849.8
Date:                Tue, 19 Mar 2024   Prob (F-statistic):           3.50e-32
Time:                        16:55:09   Log-Likelihood:                -527.44
No. Observations:                  50   AIC:                             1059.
Df Residuals:                      48   BIC:                             1063.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   4.903e+04   2537.897     19.320      0.000    4.39e+04    5.41e+04
RD             0.8543      0.029     29.151      0.000       0.795       0.913
==============================================================================
Omnibus:                       13.727   Durbin-Watson:                   1.116
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               18.536
Skew:                          -0.911   Prob(JB):                     9.44e-05
Kurtosis:                       5.361   Cond. No.                     1.65e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.65e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### RD = RD mempengaruhi profit (alpha 0,1)
#### ADJ R^2 = 0.945
#### AIC = 1059

In [ ]:
model7 = smf.ols('Profit~Administration',data = startup).fit()
model7.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Profit   R-squared:                       0.040
Model:                            OLS   Adj. R-squared:                  0.020
Method:                 Least Squares   F-statistic:                     2.015
Date:                Tue, 19 Mar 2024   Prob (F-statistic):              0.162
Time:                        16:55:09   Log-Likelihood:                -599.63
No. Observations:                  50   AIC:                             1203.
Df Residuals:                      48   BIC:                             1207.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept       7.697e+04   2.53e+04      3.040      0.004    2.61e+04    1.28e+05
Administration     0.2887      0.203      1.419      0.162      -0.120       0.698
==============================================================================
Omnibus:                        0.126   Durbin-Watson:                   0.099
Prob(Omnibus):                  0.939   Jarque-Bera (JB):                0.110
Skew:                           0.093   Prob(JB):                        0.947
Kurtosis:                       2.866   Cond. No.                     5.59e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.59e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### Administration tidak mempengaruhi profit (>0,1)

In [ ]:
model8 = smf.ols('Profit~Marketing',data = startup).fit()
model8.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Profit   R-squared:                       0.559
Model:                            OLS   Adj. R-squared:                  0.550
Method:                 Least Squares   F-statistic:                     60.88
Date:                Tue, 19 Mar 2024   Prob (F-statistic):           4.38e-10
Time:                        16:55:09   Log-Likelihood:                -580.18
No. Observations:                  50   AIC:                             1164.
Df Residuals:                      48   BIC:                             1168.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept       6e+04   7684.530      7.808      0.000    4.46e+04    7.55e+04
Marketing      0.2465      0.032      7.803      0.000       0.183       0.310
==============================================================================
Omnibus:                        4.420   Durbin-Watson:                   1.178
Prob(Omnibus):                  0.110   Jarque-Bera (JB):                3.882
Skew:                          -0.336   Prob(JB):                        0.144
Kurtosis:                       4.188   Cond. No.                     4.89e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.89e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### Marketing mempengaruhi profit (<0,1)
#### ADJ R^2 = 0.550
#### AIC = 1164

In [ ]:
model9 = smf.ols('Profit~Administration+Marketing',data = startup).fit()
model9.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Profit   R-squared:                       0.610
Model:                            OLS   Adj. R-squared:                  0.593
Method:                 Least Squares   F-statistic:                     36.71
Date:                Tue, 19 Mar 2024   Prob (F-statistic):           2.50e-10
Time:                        16:55:10   Log-Likelihood:                -577.13
No. Observations:                  50   AIC:                             1160.
Df Residuals:                      47   BIC:                             1166.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept       2.022e+04   1.77e+04      1.143      0.259   -1.54e+04    5.58e+04
Administration     0.3237      0.131      2.468      0.017       0.060       0.588
Marketing          0.2488      0.030      8.281      0.000       0.188       0.309
==============================================================================
Omnibus:                        6.584   Durbin-Watson:                   1.279
Prob(Omnibus):                  0.037   Jarque-Bera (JB):                6.524
Skew:                          -0.512   Prob(JB):                       0.0383
Kurtosis:                       4.443   Cond. No.                     1.30e+06
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.3e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### Administration mempengaruhi profit(<0,1)
#### Marketing mempengaruhi profit(<0,1)
#### ADJ R^2 = 0.593
#### AIC = 1160